File: CodingInterviewsPythonSolutions/Next Greater Element II.py
class Solution:
    def nextGreaterElements(self, nums: List[int]) -> List[int]:
        n = len(nums)
        res = [-1] * n
        stack = []
        
        for i in range(2 * n):
            while stack and nums[i % n] > nums[stack[-1]]:
                index = stack.pop()
                res[index] = nums[i % n]
            
            if i < n:
                stack.append(i)
        
        return res

File: CodingInterviewsJavaSolutions/Next Greater Element II.java
class Solution {
    public int[] nextGreaterElements(int[] nums) {
        int n = nums.length;
        int[] res = new int[n];
        Arrays.fill(res, -1);
        Deque<Integer> stack = new ArrayDeque<>();
        
        for (int i = 0; i < 2 * n; i++) {
            while (!stack.isEmpty() && nums[i % n] > nums[stack.peekLast()]) {
                int index = stack.removeLast();
                res[index] = nums[i % n];
            }
            
            if (i < n) {
                stack.addLast(i);
            }
        }
        return res;
    }
}
